### PR TITLE
Expose more Adventure serializers through PaperComponents

### PR DESCRIPTION
**Pull request by jpenilla**
Not 100% sure whether we want to expose a color downsampling gson serializer or not. My thinking behind exposing it is that if you are serializing legacy hover events, you probably don't want rgb colors in the serialized component, because you are probably sending the json to a legacy client or something.

### CODE REVIEW

File: Spigot-API-Patches/0005-Adventure.patch [hunk 1]
```diff
@@ -399,8 +399,80 @@ index 0000000000000000000000000000000000000000..13c5df5fb8ce1d0203d99e88dd691019
...
++ * Paper API-specific methods for working with {@link Component}s and related.
++ */
++public interface PaperComponents {
++    /**
++     * Get the {@link PaperComponents} instance.
++     *
++     * @return PaperComponents
++     */
++    static @NonNull PaperComponents paperComponents() {
++        return Bukkit.getPaperComponents();
```
Comment (kashike) [inline Spigot-API-Patches/0005-Adventure.patch]: Should be fetched through `Bukkit.getUnsafe()` instead.

File: Spigot-API-Patches/0005-Adventure.patch [hunk 2]
```diff
@@ -399,8 +399,80 @@ index 0000000000000000000000000000000000000000..13c5df5fb8ce1d0203d99e88dd691019
...
++     * standard JSON serialization format using Gson, downsampling any RGB colors
++     * to their nearest {@link NamedTextColor} counterpart.
++     *
++     * <p>Implementations may provide a serializer capable of processing any
++     * information that requires implementation details, such as legacy
++     * (pre-1.16) hover events.</p>
++     *
++     * @return a json component serializer
++     */
++    @NonNull GsonComponentSerializer colorDownsamplingGsonSerializer();
```
Comment (kashike) [inline Spigot-API-Patches/0005-Adventure.patch]: These should be made `static`, and have a separate ´Providers` interface for these non-`static` ones.

File: Spigot-Server-Patches/0010-Adventure.patch [hunk 3]
```diff
@@ -789,6 +789,48 @@ index 0000000000000000000000000000000000000000..71938ad7b3494e803beca7e4022aad12
...
++import io.papermc.paper.text.PaperComponents;
++import net.kyori.adventure.text.flattener.ComponentFlattener;
++import net.kyori.adventure.text.serializer.gson.GsonComponentSerializer;
++import net.kyori.adventure.text.serializer.plain.PlainComponentSerializer;
++import org.checkerframework.checker.nullness.qual.NonNull;
++
++public enum PaperComponentsImpl implements PaperComponents {
++    INSTANCE;
++
++    private final GsonComponentSerializer colorDownsamplingGsonSerializer = GsonComponentSerializer.builder()
```
Comment (kashike) [inline Spigot-Server-Patches/0010-Adventure.patch]: Should be stored in `PaperAdventure` too.

**Review by kashike — CHANGES_REQUESTED** [review]

**Review by kashike — APPROVED** [review]

File: Spigot-API-Patches/0005-Adventure.patch [hunk 4]
```diff
@@ -399,6 +399,96 @@ index 0000000000000000000000000000000000000000..13c5df5fb8ce1d0203d99e88dd691019
...
++import net.kyori.adventure.text.serializer.gson.GsonComponentSerializer;
++import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer;
++import net.kyori.adventure.text.serializer.plain.PlainComponentSerializer;
++import org.bukkit.Bukkit;
++import org.checkerframework.checker.nullness.qual.NonNull;
++
++/**
++ * Paper API-specific methods for working with {@link Component}s and related.
++ */
++public interface PaperComponents {
```
Comment (Proximyst) [inline Spigot-API-Patches/0005-Adventure.patch]: This only contains static methods and has no implementation. Perhaps worth preferring an empty enum or a non-instantiable class for it instead?

**Review by Proximyst — DISMISSED** [review]

File: Spigot-API-Patches/0005-Adventure.patch (hunk 4 above)
Comment (kashike) [reply Spigot-API-Patches/0005-Adventure.patch]: `final class` with `private` constructor.

Comment (jpenilla) [reply Spigot-API-Patches/0005-Adventure.patch]: Updated

**Review by zml2008 — APPROVED** [review]

**Pull request merged**